src/main/scala/com/datastax/gatling/plugin/request/CqlRequestAction.scala

```diff
@@ -61,48 +62,53 @@ class CqlRequestAction(val name: String,
     })
   }
 
-  def sendQuery(session: Session): Unit = {
-    val enableCO = Boolean.getBoolean("gatling.dse.plugin.measure_service_time")
-    val responseTimeBuilder: ResponseTimeBuilder = if (enableCO) {
-      // The throughput checker is useless in CO affected scenarios since throughput is not known in advance
-      COAffectedResponseTime.startingNow(gatlingTimingSource)
-    } else {
-      ThroughputVerifier.checkForGatlingOverloading(session, gatlingTimingSource)
-      GatlingResponseTime.startedByGatling(session, gatlingTimingSource)
-    }
-    val stmt = safely()(dseAttributes.statement.buildFromSession(session))
+  private def buildStatement(builder:B):T = {
 
-    stmt.onFailure(err => {
-      handleFailure(session, responseTimeBuilder, err)
-    })
+    // global options
+    dseAttributes.cl.foreach(builder.setConsistencyLevel)
+    dseAttributes.idempotent.foreach(builder.setIdempotence(_))
+    dseAttributes.node.foreach(builder.setNode)
 
-    stmt.onSuccess({ stmt =>
-      // global options
-      dseAttributes.cl.map(stmt.setConsistencyLevel)
-      dseAttributes.userOrRole.map(stmt.executingAs)
-      dseAttributes.readTimeout.map(stmt.setReadTimeoutMillis)
-      dseAttributes.idempotent.map(stmt.setIdempotent)
-      dseAttributes.defaultTimestamp.map(stmt.setDefaultTimestamp)
-
-      // CQL Only Options
-      dseAttributes.outGoingPayload.map(x => stmt.setOutgoingPayload(x.asJava))
-      dseAttributes.serialCl.map(stmt.setSerialConsistencyLevel)
-      dseAttributes.retryPolicy.map(stmt.setRetryPolicy)
-      dseAttributes.fetchSize.map(stmt.setFetchSize)
-      dseAttributes.pagingState.map(stmt.setPagingState)
-      if (dseAttributes.enableTrace.isDefined && dseAttributes.enableTrace.get) {
-        stmt.enableTracing
+    // CQL Only Options
+    dseAttributes.customPayload.foreach {
+      _.foreach {
+        _ match {
+          case (k, v) => builder.addCustomPayload(k, v)
+        }
       }
+    }
+    dseAttributes.enableTrace.foreach(builder.setTracing)
+    dseAttributes.pageSize.foreach(builder.setPageSize)
+    dseAttributes.pagingState.foreach(builder.setPagingState)
+    dseAttributes.queryTimestamp.foreach(builder.setQueryTimestamp)
+    dseAttributes.routingKey.foreach(builder.setRoutingKey)
+    dseAttributes.routingKeyspace.foreach(builder.setRoutingKeyspace)
+    dseAttributes.routingToken.foreach(builder.setRoutingToken)
+    dseAttributes.serialCl.foreach(builder.setSerialConsistencyLevel)
+    dseAttributes.timeout.foreach(builder.setTimeout)
+    builder.build
+  }
 
-      val responseHandler = new CqlResponseHandler(next, session, system, statsEngine, responseTimeBuilder, stmt, dseAttributes, metricsLogger)
-      implicit val sameThreadExecutionContext: ExecutionContextExecutor = ExecutionContext.fromExecutorService(dseExecutorService)
-      FutureUtils
-        .toScalaFuture(protocol.session.executeAsync(stmt))
```

Comment:
why we did resign from `toScalaFuture`?
I think we could create a similar method for CompletableFuture -> Future (Scala version)
The scala `Future` has a rich API.

Reply:
We still convert the Java future to a Scala future here, it's just done elsewhere.  As part of my changes I tried to break success and failure handlers into distinct fns (handleSuccess and handleFailure respectively).  This class wasn't doing that; when I got to it the handling fns were declared inline in sendQuery() and then added manually.

If you look at the end of handleSuccess you see it's still there:

```scala
  private def handleSuccess(session: Session, responseTimeBuilder: ResponseTimeBuilder)(builder:B): Unit = {

    val stmt:T = buildStatement(builder)
    val responseHandler =
      new CqlResponseHandler[T, B](
        next,
        session,
        system,
        statsEngine,
        responseTimeBuilder,
        stmt,
        dseAttributes,
        metricsLogger)
    implicit val sameThreadExecutionContext: ExecutionContextExecutor = ExecutionContext.fromExecutorService(dseExecutorService)
    FutureConverters
      .toScala(protocol.session.executeAsync(stmt))
      .onComplete(t => DseRequestActor.recordResult(RecordResult(t, responseHandler)))
  }
```